Subject: mysys/get_password.c: read passwords without echo on OpenVMS

As patch 0011 for Connector/C: there is no tcgetattr()/tcsetattr() in the
OpenVMS C RTL (link: undefined symbols in every client). Read from
SYS$COMMAND with a no-echo $QIOW, or a line from stdin without a terminal.
buff[] is on the stack, in the 32-bit address space every terminal driver
accepts.

--- a/mysys/get_password.c	2026-10-04 14:16:07.784416207 +0100
+++ b/mysys/get_password.c	2026-10-04 14:16:07.804307164 +0100
@@ -52,6 +52,12 @@
 #endif /* _WIN32 */
 #endif /* HAVE_GETPASS */
 
+#ifdef __VMS
+#include <descrip.h>
+#include <iodef.h>
+#include <starlet.h>
+#endif
+
 #ifdef HAVE_GETPASSPHRASE			/* For Solaris */
 #define getpass(A) getpassphrase(A)
 #endif
@@ -193,7 +199,39 @@
     fputs(opt_message ? opt_message : "Enter password: ",stderr);
     fflush(stderr);
   }
-#if defined(HAVE_TERMIOS_H)
+#if defined(__VMS)
+  /*
+    No tcgetattr()/tcsetattr() in the OpenVMS C RTL: read from the terminal
+    with a no-echo QIO (as Connector/C's get_tty_password() does there), or a
+    line from stdin when there is no terminal.
+  */
+  {
+    $DESCRIPTOR(tt, "SYS$COMMAND");
+    unsigned short chan;
+    struct { unsigned short status, count; unsigned int info; } iosb;
+    int sts= sys$assign(&tt, &chan, 0, 0);
+    buff[0]= 0;
+    if (sts & 1)
+    {
+      sts= sys$qiow(0, chan, IO$_READVBLK | IO$M_NOECHO, &iosb, 0, 0,
+                    buff, sizeof(buff) - 1, 0, 0, 0, 0);
+      if (sts & 1)
+        sts= iosb.status;
+      sys$dassgn(chan);
+      if (sts & 1)
+        buff[iosb.count < sizeof(buff) - 1 ? iosb.count : sizeof(buff) - 1]= 0;
+    }
+    if (!(sts & 1))
+    {
+      size_t n;
+      if (!fgets(buff, sizeof(buff), stdin))
+        buff[0]= 0;
+      n= strlen(buff);
+      while (n && (buff[n - 1] == '\n' || buff[n - 1] == '\r'))
+        buff[--n]= 0;
+    }
+  }
+#elif defined(HAVE_TERMIOS_H)
   tcgetattr(fileno(stdin), &org);
   tmp = org;
   tmp.c_lflag &= ~(ECHO | ISIG | ICANON);
